documentation/flows/agent/dynamic_backhaul_switching.md

```diff
@@ -0,0 +1,113 @@
+Future enhancement for selecting the better link (based on phy rate) is out of scope of this document.
+3.	UCC commands – for certification, dev_reset_default should reset any previous onboarding state. This means disabling the backhaul, whichever it is. Later, when a dev_set_config is received, switching to the requested backhaul type will take place.
+
+## Current state
+
+Currently, there is no way to dynamically switch from wired to wireless backhaul and vice versa.
+There is a disabled polling-based mechanism which switches to the wired backhaul if it is up, but it is disabled since it consumed a lot of CPU. 
+Specifically, for certification, only wired backhaul onboarding is supported – e.g. only wired dev_set_config is supported now.
+
+This design provides the guidelines to create a generic mechanism at the backhaul manager to provide backhaul switching support. It will provide an API at the backhaul manager to switch backhaul and start onboarding, which can be triggered from either a UCC command, link up/down events or from a controller backhaul steering request.
+
+The following state diagram shows the backhaul FSM high level design:
+
+![UML](../../images/plantuml/agent/state_machines/backhaul_manager_fsm.png)
+
+The above figure shows that currently, there is a simple FSM restart on wired link status change (which is commented out in the code). This mechanism is currently disabled since wireless-backhaul is not yet  supported and the current implementation causes high CPU load on steady-state.
+
+---
+**NOTE**
+
+ The Wireless FSM state is out of scope for this design, since backhaul switching takes place in OPERATIONAL state alone - once the FSM is at the OPERATIONAL state, backhaul switching can be triggered in the backhaul manager on demand, and will result with FSM restart, which in turn will result with selecting a new backhaul.
+
+---
+
+## Requirements
+
+* Generic implementation in the backhaul manager which will provide a single entrypoint to perform backhaul switching from multiple clients
+* Supported clients - UCC, Controller, netlink events
+* Prevent loops - use only one backhaul at a time
+* Repeater static 
+
+## Assumptions
+
+* Wireless backhaul support for certification exists, including `dev_set_config,name,DUT,program,map,backhaul,RUID`
+* Repeater static configuration
+  * hostapds are configured with start_disabled=1 and static channel at boot
+    * the effect is that the AP will NOT hold a channel and will not put anything on-the-air
+  * Supplicants are configured for WPS_PBC - `wpa_cli wps_pbc multi_ap=1
+` in onboarding mode
+  * Supplicants are configured with credentials in operational mode
+
+# High Level Design
+
+Dynamic backhaul switching will be implemented as an API in the backhaul manager - `backhaul_manager_thread::switch_backhaul(const std::string iface_name)` which will move the backhaul manager FSM to the RESTART state, which also causes both radio agents restart.
```

Comment:
I believe (@morantr please confirm) that after unified agent, the backhaul's RESTART state will no longer cause the ap_managers to restart. Which is a good thing IMO. But I don't think it's worth mentioning here either way.

Reply:
It depends on what we decide that will happen.
Personally, I think that if the backhaul is disconnected and there is no link backward, there is no reason to allow clients to remain connected since there they will not get service for who knows how much time.

Reply:
> Personally, I think that if the backhaul is disconnected (because of steering request or any other reason) and there is no link backward, there is no reason to allow clients to remain connected since there they will not get service for who knows how much time.

If we do it well, the time that it remains disconnected can be small. Also, if we bring down the AP, most clients will try to connect to another AP, which is usually not what we want.

> Also, when we need to establish a wireless link, we need to scan the air first, which means that the radio channel will have to change, so it is better if the AP will be down at that moment.

Do we actually need to scan? Typically the APs don't change very much, so we can cache scan results for a very long time. Also, I think in many situations all the bAPs will be on the same channel (https://github.com/prplfoundation/prplMesh/pull/1423#discussion_r444854037) so scan/probe can easily be done while operating the AP. Also, at least some drivers do support off-channel scanning while operating an AP (at least I think so). And don't forget the situation where there's a dedicated radio for backhaul - then backhaul steering without interruption is definitely possible. Also if we switch radios. And finally, because the steering request contains the target channel and we have the BSSID, it's not even necessary to scan, we can directly send a probe request to that bssid (though I don't think this is supported by wpa_supplicant at the moment).